File: quam_state_manager/web/static/search-query.js
/* SearchQuery — THE one boolean structure for SM's search boxes.
 *
 * The app grew five mutually incompatible search semantics (whole-substring
 * trees, three different AND tokenizers, two scoped grammars), and users kept
 * asking why two words work in Live State Edit but not in the Json Tree View.
 * This module owns the part every surface shares — tokenization for the plain
 * surfaces, and the AND/OR group structure for all of them — so they cannot
 * drift apart again. (The failure shape is proven, twice over: the two
 * row-hide classes of which only one was filtered, and the two grid
 * addressing models — docs/62. Even the two scoped grammars that CLAIM to
 * mirror each other have drifted on `-x=y` and on comma splitting.)
 *
 * Grammar:
 *
 *   space          AND — every group must match. For a query of plain words
 *                  this module produces exactly `split(/\s+/)` as singleton
 *                  groups, so every surface's historic behaviour is the
 *                  no-pipe special case, character for character.
 *
 *   " | "          OR — a STANDALONE pipe token with a joinable term on BOTH
 *                  sides merges its neighbours into one alternation group.
 *                  Alternation binds TIGHTER than conjunction (as in Google):
 *                  `x180 amplitude | length` = x180 AND (amplitude|length).
 *                  Any pipe that is not exactly that — embedded (`|e>`),
 *                  leading, trailing, doubled, or beside a negated term —
 *                  stays a LITERAL term, which is what it was yesterday.
 *
 * The standalone-token guard is measured, not stylistic: `|` occurs in 0 of
 * 43,115 distinct terms across 41 real chip states, 0 of the whole Datasets
 * haystack and 0 of 102 grid column labels — but node.json description
 * strings carry ket notation (`|e>`, `|00>`, `|baseline|`) in 25.7% of 2,829
 * real run files, and the dataset detail tab renders those through the same
 * tree search. An `OR` keyword was rejected on the same evidence: 21.9% of
 * tree nodes contain "or" (resonator, readout, DragCosine...).
 *
 * Surfaces with richer grammars (the dataset table's scopes and negation, the
 * sidebar tree filter's server-side twin in core/search_query.py) keep their
 * own token CLASSIFIERS and compose here at the group level via groupBy():
 * split into OR groups, run the existing per-token semantics inside each.
 * core/search_query.py mirrors this file and is pinned to it by
 * tests/test_search_query.py — change BOTH or the parity test fails.
 */
window.SearchQuery = (function () {
    'use strict';

    /** Lowercased whitespace tokens; [] for an empty/blank query. */
    function tokens(q) {
        q = (q == null ? '' : String(q)).toLowerCase().trim();
        return q ? q.split(/\s+/) : [];
    }

    /**
     * Group ANY ordered token list into AND-of-OR groups (tight-binding `|`).
     *
     * `list` items can be plain strings or a surface's structured tokens;
     * `getRaw(item)` must return the raw token string (identity by default)
     * — only an item whose raw is exactly "|" can be an operator; and
     * `joinable(item)` says whether an item may be an OR operand (default
     * yes; scoped surfaces pass `not negated`, so `-tag:wip | x` keeps its
     * pipe literal instead of inventing OR-of-negation semantics).
     *
     * Returns an array of groups; each group is an array of items, meaning
     * (AND over groups) of (OR within a group). Every non-operator pipe is
     * its own literal group — i.e. exactly the term it always was.
     */
    function groupBy(list, getRaw, joinable) {
        getRaw = getRaw || function (x) { return x; };
        joinable = joinable || function () { return true; };
        var groups = [];        // parallel: canJoin[i] = last item of groups[i] joinable
        var canJoin = [];
        for (var i = 0; i < list.length; i++) {
            var item = list[i];
            if (getRaw(item) === '|') {
                var nxt = (i + 1 < list.length) ? list[i + 1] : null;
                if (groups.length && canJoin[groups.length - 1] &&
                        nxt !== null && getRaw(nxt) !== '|' && joinable(nxt)) {
                    groups[groups.length - 1].push(nxt);   // a | b  →  {a, b}
                    canJoin[groups.length - 1] = true;
                    i++;                                    // consumed the operand
                    continue;
                }
                groups.push([item]);                        // literal pipe
                canJoin.push(false);
                continue;
            }
            groups.push([item]);
            canJoin.push(!!joinable(item));
        }
        return groups;
    }

    /** Convenience for plain-string surfaces: query → groups of strings. */
    function groups(q) {
        return groupBy(tokens(q));
    }

    /** (AND over groups) of (OR within a group) of substring tests. */
    function matchesHay(hayLower, grps) {
        for (var g = 0; g < grps.length; g++) {
            var toks = grps[g], any = false;
            for (var t = 0; t < toks.length; t++) {
                if (hayLower.indexOf(toks[t]) >= 0) { any = true; break; }
            }
            if (!any) return false;
        }
        return true;
    }

    /* The same character class Python's str.isspace() uses for the characters
       a query can actually carry. NOT /\s/: Python calls U+0085 whitespace and
       JS does not, so a regex here would split a pasted query differently on
       the two sides of the wire. */
    function _isSep(ch) {
        return ch === ' ' || ch === ',' || ch === '\t' || ch === '\n'
            || ch === '\r' || ch === '\f' || ch === '\v'
            || ch === '\u001c' || ch === '\u001d' || ch === '\u001e'
            || ch === '\u001f' || ch === '\u0085' || ch === '\u00a0'
            || ch === '\u1680' || (ch >= '\u2000' && ch <= '\u200a')
            || ch === '\u2028' || ch === '\u2029' || ch === '\u202f'
            || ch === '\u205f' || ch === '\u3000';
    }

    /* Every token as [rawStart, rawEnd, text] — the twin of
       core.search_query.scoped_spans. The span is RAW (it covers the quote
       characters the text drops), because a caller splicing into an input's
       value needs offsets into what the user can see. */
    function scopedSpans(text) {
        var out = [], cur = '', start = -1, inQ = false;
        text = String(text == null ? '' : text);
        for (var i = 0; i < text.length; i++) {
            var ch = text.charAt(i);
            if (ch === '"') {
                if (start < 0) start = i;
                inQ = !inQ;
                continue;
            }
            if (!inQ && _isSep(ch)) {
                if (cur) out.push([start, i, cur]);
                cur = ''; start = -1;
                continue;
            }
            if (start < 0) start = i;
            cur += ch;
        }
        if (cur) out.push([start, text.length, cur]);
        return out;
    }

    function scopedTokens(text) {
        return scopedSpans(text).map(function (s) { return s[2]; });
    }

    /* The token the caret is FINISHING, or null. Refuses the two states where
       completing would move text the user did not mean: a token not begun (the
       caret at 0 or straight after a separator), and an unbalanced quote (past
       an open quote the scan finds no separator, so the span runs to the end of
       the box). */
    function caretSpan(text, caret) {
        text = String(text == null ? '' : text);
        if (caret == null || caret <= 0 || caret > text.length) return null;
        var q = 0;
        for (var i = 0; i < caret; i++) if (text.charAt(i) === '"') q++;
        if (q % 2) return null;
        // No separator test: `spans[j][0] < caret` below already excludes a
        // caret on one. Measured redundant over 219,344 (text, caret) pairs.
        var spans = scopedSpans(text);
        for (var j = 0; j < spans.length; j++) {
            if (spans[j][0] < caret && caret <= spans[j][1]) return spans[j];
        }
        return null;
    }

    return { tokens: tokens, groupBy: groupBy, groups: groups,
             matchesHay: matchesHay,
             scopedSpans: scopedSpans, scopedTokens: scopedTokens,
             caretSpan: caretSpan };
})();

File: quam_state_manager/core/search_query.py
"""The server-side twin of ``web/static/search-query.js`` — ONE query grammar.

The sidebar workspace filter runs on the server (``routes._parse_tree_query``)
while every other search box runs in the browser, and the two scoped grammars
that claimed to mirror each other had already drifted (``-x=y`` negates on one
and is literal on the other; commas split on one and not the other). This
module and its JS twin carry the SHARED part — plain tokenization and the
AND-of-OR group structure — and are pinned to each other case-for-case by
``tests/test_search_query.py`` (the ``value_delta`` precedent: two languages,
one behaviour, a test that fails when either side moves alone).

Grammar (see the JS twin's header for the measurements behind it):

* whitespace = AND — a plain-word query becomes singleton groups, i.e. the
  historic behaviour character for character;
* a STANDALONE ``|`` token with a joinable term on both sides merges its
  neighbours into one OR group, binding tighter than AND
  (``x180 amplitude | length`` = x180 AND (amplitude OR length));
* every other pipe — embedded (``|e>`` ket notation), leading, trailing,
  doubled, or beside a negated term — stays a literal term.
"""

from __future__ import annotations

from typing import Any, Callable, Sequence


def tokens(q: str | None) -> list[str]:
    """Lowercased whitespace tokens; ``[]`` for an empty/blank query."""
    q = ("" if q is None else str(q)).lower().strip()
    return q.split() if q else []


def group_by(
    items: Sequence[Any],
    get_raw: Callable[[Any], str] | None = None,
    joinable: Callable[[Any], bool] | None = None,
) -> list[list[Any]]:
    """Group ANY ordered token list into AND-of-OR groups (tight-binding ``|``).

    ``get_raw(item)`` must return the raw token string (identity by default) —
    only an item whose raw is exactly ``"|"`` can be an operator. ``joinable``
    says whether an item may be an OR operand (default yes; scoped surfaces
    pass ``not negated``). Returns groups of items: AND over groups, OR within
    a group. Every non-operator pipe is its own literal group.
    """
    if get_raw is None:
        get_raw = lambda x: x  # noqa: E731 — mirrors the JS default exactly
    if joinable is None:
        joinable = lambda x: True  # noqa: E731

    groups: list[list[Any]] = []
    can_join: list[bool] = []
    i = 0
    n = len(items)
    while i < n:
        item = items[i]
        if get_raw(item) == "|":
            nxt = items[i + 1] if i + 1 < n else None
            if (groups and can_join[-1] and nxt is not None
                    and get_raw(nxt) != "|" and joinable(nxt)):
                groups[-1].append(nxt)          # a | b  ->  {a, b}
                can_join[-1] = True
                i += 2                          # consumed the operand
                continue
            groups.append([item])               # literal pipe
            can_join.append(False)
            i += 1
            continue
        groups.append([item])
        can_join.append(bool(joinable(item)))
        i += 1
    return groups


def groups(q: str | None) -> list[list[str]]:
    """Convenience for plain-string surfaces: query -> groups of strings."""
    return group_by(tokens(q))


def matches_hay(hay_lower: str, grps: Sequence[Sequence[str]]) -> bool:
    """(AND over groups) of (OR within a group) of substring tests."""
    for grp in grps:
        if not any(tok in hay_lower for tok in grp):
            return False
    return True


# ── The one search hint, docs/141 4aj (user-directed) ────────────────────────
#
# Thirteen search boxes carried thirteen hand-written placeholders ("Search
# keys or values...", "Search all pulses…", "Search: q2, q5, time · rabi ·
# tag:flagged · is:bookmarked"), and none of them said what the grammar
# actually is — so the AND/OR the whole app shares was invisible everywhere
# except the two boxes that happened to spell it out. The grammar IS the
# thing worth saying in the little space a placeholder has: examples are
# guessable, operators are not.
#
# Shape: "Search: space = AND, | = OR" plus, only where the surface really
# has them, its own scope tokens. The full sentence lives in the `title`
# (SEARCH_TITLE) so nothing is lost to the compaction.
HINT = "space = AND, | = OR"
SEARCH_TITLE = (
    "Space between words = AND (every word must match). "
    "A standalone | between two words = OR. Any other pipe is a literal."
)


def search_hint(*extras: str) -> str:
    """The placeholder every SM search box uses.

    ``extras`` are the surface's OWN scope tokens (``"tag:"``, ``"is:"``) —
    kept to what that box can really do, appended after the grammar and
    never instead of it.
    """
    parts = [HINT] + [e for e in extras if e]
    return "Search: " + ", ".join(parts)


def search_title(*extras: str) -> str:
    """The full grammar sentence for the box's tooltip."""
    tail = (" Scopes here: " + ", ".join(e for e in extras if e) + ".") if any(extras) else ""
    return SEARCH_TITLE + tail
